File: scripts/settle.lua
-- EchoQueue settle. The receipt is the first-terminal-command fence.
-- KEYS[1] pending
-- KEYS[2] receipt
-- KEYS[3] namespace deadline zset
-- KEYS[4] result list
-- KEYS[5] source list for immediate retries
-- KEYS[6] dead list
-- ARGV[1] batch id
-- ARGV[2] request id
-- ARGV[3] trusted command hash
-- ARGV[4] receipt base JSON
-- ARGV[5] result records JSON
-- ARGV[6] retry tasks JSON
-- ARGV[7] dead records JSON
-- ARGV[8] receipt TTL in milliseconds

local pendingKey = KEYS[1]
local receiptKey = KEYS[2]
local deadlineKey = KEYS[3]
local resultKey = KEYS[4]
local sourceKey = KEYS[5]
local deadKey = KEYS[6]
local batchID = ARGV[1]
local requestID = ARGV[2]
local commandHash = ARGV[3]
local receiptTTL = tonumber(ARGV[8])

if not receiptTTL or receiptTTL < 1 then
    return {"invalid", "receipt ttl is invalid"}
end

local function validateReceipt(raw, expectedBatchID, expectedRequestID, expectedHash, expectedWinner, requireClosed)
    local ok, receipt = pcall(cjson.decode, raw)
    if not ok or type(receipt) ~= 'table' then
        return nil, "receipt is corrupt"
    end
    if receipt.schema_version ~= 1 or receipt.protocol_version ~= 1 then
        return nil, "receipt protocol is invalid"
    end
    if type(receipt.batch_id) ~= 'string' or receipt.batch_id ~= expectedBatchID then
        return nil, "receipt batch id is invalid"
    end
    if type(receipt.request_id) ~= 'string' or receipt.request_id == '' then
        return nil, "receipt request id is invalid"
    end
    if expectedRequestID and receipt.request_id ~= expectedRequestID then
        return nil, "receipt request id does not match command"
    end
    if type(receipt.command_hash) ~= 'string' or receipt.command_hash == '' then
        return nil, "receipt command hash is invalid"
    end
    if expectedHash and receipt.command_hash ~= expectedHash then
        return nil, "receipt command hash does not match command"
    end
    if expectedWinner then
        if receipt.winner ~= expectedWinner then
            return nil, "receipt winner is invalid"
        end
    elseif receipt.winner ~= 'settle' and receipt.winner ~= 'recover' then
        return nil, "receipt winner is invalid"
    end
    if type(receipt.closed_at) ~= 'number' or (requireClosed and receipt.closed_at <= 0) or (not requireClosed and receipt.closed_at < 0) then
        return nil, "receipt closed_at is invalid"
    end
    for _, field in ipairs({"result_count", "retry_count", "dead_count"}) do
        if type(receipt[field]) ~= 'number' or receipt[field] < 0 or receipt[field] % 1 ~= 0 then
            return nil, "receipt count is invalid"
        end
    end
    return receipt, nil
end

local receiptType = redis.call('TYPE', receiptKey)['ok']
if receiptType ~= 'none' and receiptType ~= 'string' then
    return {"invalid", "receipt key has wrong type"}
end
local existing = nil
if receiptType == 'string' then
    existing = redis.call('GET', receiptKey)
end
if existing then
    local parsed, receiptErr = validateReceipt(existing, batchID, nil, nil, nil, true)
    if not parsed then return {"invalid", receiptErr} end
    if parsed.request_id == requestID then
        if parsed.command_hash == commandHash then
            return {"duplicate", existing}
        end
        return {"conflict", existing}
    end
    return {"stale", existing}
end

local pendingRaw = redis.call('GET', pendingKey)
if not pendingRaw then
    return {"not_found"}
end
local pendingOK, pending = pcall(cjson.decode, pendingRaw)
if not pendingOK or type(pending) ~= 'table' then
    return {"invalid", "pending is corrupt"}
end
if pending.schema_version ~= 1 or pending.protocol_version ~= 1 or pending.state ~= "PENDING" or pending.batch_id ~= batchID then
    return {"invalid", "pending is not the requested open batch"}
end

local deadlineType = redis.call('TYPE', deadlineKey)['ok']
if deadlineType ~= 'none' and deadlineType ~= 'zset' then
    return {"invalid", "deadline key has wrong type"}
end

local resultOK, results = pcall(cjson.decode, ARGV[5])
local retryOK, retries = pcall(cjson.decode, ARGV[6])
local deadOK, dead = pcall(cjson.decode, ARGV[7])
if not resultOK or type(results) ~= 'table' or not retryOK or type(retries) ~= 'table' or not deadOK or type(dead) ~= 'table' then
    return {"invalid", "effect records are corrupt"}
end

local function checkList(key, needed)
    if needed == 0 then return true end
    local kind = redis.call('TYPE', key)['ok']
    return kind == 'none' or kind == 'list'
end
if not checkList(resultKey, #results) then return {"invalid", "result key has wrong type"} end
if not checkList(sourceKey, #retries) then return {"invalid", "source key has wrong type"} end
if not checkList(deadKey, #dead) then return {"invalid", "dead key has wrong type"} end

for _, record in ipairs(results) do
    if type(record) ~= 'table' or record.schema_version ~= 1 or record.protocol_version ~= 1 or not record.effect_id or not record.task_id then
        return {"invalid", "result record identity is missing"}
    end
end
for _, task in ipairs(retries) do
    if type(task) ~= 'table' or not task.task_id or task.retry_count == nil or task.payload == nil then
        return {"invalid", "retry task is malformed"}
    end
end
for _, record in ipairs(dead) do
    if type(record) ~= 'table' or record.schema_version ~= 1 or record.protocol_version ~= 1 or not record.effect_id or not record.task_id then
        return {"invalid", "dead record identity is missing"}
    end
end

local receipt, receiptErr = validateReceipt(ARGV[4], batchID, requestID, commandHash, 'settle', false)
if not receipt then return {"invalid", receiptErr} end
if receipt.result_count ~= #results or receipt.retry_count ~= #retries or receipt.dead_count ~= #dead then
    return {"invalid", "receipt counts do not match effects"}
end

for _, record in ipairs(results) do redis.call('RPUSH', resultKey, cjson.encode(record)) end
for _, task in ipairs(retries) do redis.call('RPUSH', sourceKey, cjson.encode(task)) end
for _, record in ipairs(dead) do redis.call('RPUSH', deadKey, cjson.encode(record)) end

local now = redis.call('TIME')
receipt.closed_at = tonumber(now[1]) * 1000 + math.floor(tonumber(now[2]) / 1000)
local receiptJSON = cjson.encode(receipt)
if not redis.call('SET', receiptKey, receiptJSON, 'NX', 'PX', receiptTTL) then
    local raced = redis.call('GET', receiptKey)
    if raced then return {"stale", raced} end
    return {"invalid", "receipt could not be stored"}
end

redis.call('ZREM', deadlineKey, batchID)
redis.call('DEL', pendingKey)
return {"applied", receiptJSON}

File: settle.go
package echoqueue

import (
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"sort"

	"github.com/redis/go-redis/v9"
)

//go:embed scripts/settle.lua
var settleScriptSource string

// settleScript runs the embedded script through EVALSHA with an automatic
// EVAL fallback, so the hot path never resends the script body.
var settleScript = redis.NewScript(settleScriptSource)

var errSettle = errors.New("echoqueue: settle failed")

// errResultTooLarge is the internal, unexported marker for Settle outcomes
// whose Result data exceeds the configured byte limits. Oversized results are
// rejected before any Redis capability probe, Pending read, or Lua call, so
// Pending, deadline, Receipt, and effect lists are never touched.
var errResultTooLarge = errors.New("echoqueue: result exceeds configured size limit")

// Settle accepts a batch ID and worker outcome. The pending snapshot supplies
// the queue and policy, so callers never repeat task or routing information.
func (s *Scheduler) Settle(ctx context.Context, batchID string, outcome Outcome) (Receipt, error) {
	if s == nil || s.rdb == nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: scheduler is nil", errSettle)
	}
	if ctx == nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: context is nil", errSettle)
	}
	if err := validateText("batch_id", batchID, true); err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: %v", errSettle, err)
	}
	if err := outcome.validate(); err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: %v", errSettle, err)
	}
	if err := validateOutcomeResultSizes(s.config.MaxPayloadBytes, s.config.MaxBatchBytes, outcome); err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: %v", errSettle, err)
	}
	if err := s.ensureRedis(ctx); err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, err
	}
	command, err := commandHash(batchID, outcome)
	if err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: command hash: %v", errSettle, err)
	}

	pendingKey := s.keys.pending(batchID)
	receiptKey := s.keys.receipt(batchID)
	deadlineKey := s.keys.deadline()
	var snapshot pendingSnapshot
	pendingRaw, pendingErr := s.rdb.Get(ctx, pendingKey).Result()
	if pendingErr != nil && pendingErr != redis.Nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: read pending: %v", errSettle, pendingErr)
	}
	resultRecords := make([]resultRecord, 0)
	retryTasks := make([]Task, 0)
	deadRecords := make([]deadRecord, 0)
	if pendingErr == nil {
		if err := json.Unmarshal([]byte(pendingRaw), &snapshot); err != nil {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: decode pending: %v", errSettle, err)
		}
		if err := snapshot.validate(); err != nil {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: pending: %v", errSettle, err)
		}
		if snapshot.BatchID != batchID {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: pending batch id mismatch", errSettle)
		}
		resultRecords, retryTasks, deadRecords, err = buildOutcomeEffects(snapshot, outcome)
		if err != nil {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: %v", errSettle, err)
		}
	}

	resultKey := s.keys.result("unknown")
	sourceKey := ""
	deadKey := s.keys.deadFor("unknown")
	if pendingErr == nil {
		sourceKey = snapshot.Queue.Source
		resultKey = snapshot.Queue.Result
		if resultKey == "" {
			resultKey = s.keys.result(snapshot.Queue.TaskName)
		}
		deadKey = snapshot.Queue.Dead
		if deadKey == "" {
			deadKey = s.keys.deadFor(snapshot.Queue.TaskName)
		}
	}
	receiptJSON, err := json.Marshal(storedReceipt{
		SchemaVersion: protocolVersion,
		Protocol:      protocolVersion,
		BatchID:       batchID,
		RequestID:     outcome.RequestID,
		CommandHash:   command,
		Winner:        "settle",
		ResultCount:   len(resultRecords),
		RetryCount:    len(retryTasks),
		DeadCount:     len(deadRecords),
	})
	if err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: encode receipt: %v", errSettle, err)
	}
	resultJSON, _ := json.Marshal(resultRecords)
	retryJSON, _ := json.Marshal(retryTasks)
	deadJSON, _ := json.Marshal(deadRecords)
	value, err := settleScript.Eval(ctx, s.rdb, []string{pendingKey, receiptKey, deadlineKey, resultKey, sourceKey, deadKey},
		batchID, outcome.RequestID, command, string(receiptJSON), string(resultJSON), string(retryJSON), string(deadJSON), s.config.ReceiptTTL.Milliseconds()).Result()
	if err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: Redis script: %v", errSettle, err)
	}
	return parseReceiptResponse(batchID, value, errSettle)
}

// validateOutcomeResultSizes enforces the same byte limits Dispatch applies to
// task payloads: each Result.Data must fit within maxPayloadBytes and the sum
// of all Result.Data in the outcome must fit within maxBatchBytes. Sizes are
// measured as the raw JSON bytes of Result.Data, and the total is accumulated
// with a remaining-budget that can never overflow. Failure entries are not
// counted because they reference payloads already limited at Dispatch time.
func validateOutcomeResultSizes(maxPayloadBytes, maxBatchBytes int, outcome Outcome) error {
	remaining := maxBatchBytes
	for _, result := range outcome.Results {
		size := len(result.Data)
		if size > maxPayloadBytes {
			return fmt.Errorf("%w: result task_id %q is %d bytes, exceeds max_payload_bytes %d", errResultTooLarge, result.TaskID, size, maxPayloadBytes)
		}
		if size > remaining {
			return fmt.Errorf("%w: result data for outcome totals %d bytes, exceeds max_batch_bytes %d", errResultTooLarge, maxBatchBytes-remaining+size, maxBatchBytes)
		}
		remaining -= size
	}
	return nil
}

func buildOutcomeEffects(snapshot pendingSnapshot, outcome Outcome) ([]resultRecord, []Task, []deadRecord, error) {
	tasks := make(map[string]Task, len(snapshot.Tasks))
	for _, task := range snapshot.Tasks {
		tasks[task.TaskID] = task
	}
	seen := make(map[string]struct{}, len(outcome.Results)+len(outcome.Failures))
	results := make([]resultRecord, 0, len(outcome.Results))
	for _, result := range outcome.Results {
		task, ok := tasks[result.TaskID]
		if !ok {
			return nil, nil, nil, fmt.Errorf("unknown result task_id %q", result.TaskID)
		}
		seen[result.TaskID] = struct{}{}
		results = append(results, resultRecord{
			SchemaVersion: protocolVersion,
			Protocol:      protocolVersion,
			EffectID:      effectID("result", snapshot.BatchID, result.TaskID, task.RetryCount),
			TaskID:        result.TaskID,
			BatchID:       snapshot.BatchID,
			Data:          result.Data,
		})
	}
	retries := make([]Task, 0, len(outcome.Failures))
	dead := make([]deadRecord, 0, len(outcome.Failures))
	for _, failure := range outcome.Failures {
		task, ok := tasks[failure.TaskID]
		if !ok {
			return nil, nil, nil, fmt.Errorf("unknown failure task_id %q", failure.TaskID)
		}
		seen[failure.TaskID] = struct{}{}
		if failure.Retryable && task.RetryCount < snapshot.MaxRetry {
			task.RetryCount++
			retries = append(retries, task)
			continue
		}
		dead = append(dead, deadRecord{
			SchemaVersion: protocolVersion,
			Protocol:      protocolVersion,
			EffectID:      effectID("dead", snapshot.BatchID, failure.TaskID, task.RetryCount),
			TaskID:        failure.TaskID,
			BatchID:       snapshot.BatchID,
			RetryCount:    task.RetryCount,
			Reason:        failure.Reason,
			Payload:       task.Payload,
		})
	}
	if len(seen) != len(tasks) {
		return nil, nil, nil, fmt.Errorf("outcome covers %d of %d pending tasks", len(seen), len(tasks))
	}
	sort.Slice(results, func(i, j int) bool { return results[i].TaskID < results[j].TaskID })
	sort.Slice(retries, func(i, j int) bool { return retries[i].TaskID < retries[j].TaskID })
	sort.Slice(dead, func(i, j int) bool { return dead[i].TaskID < dead[j].TaskID })
	return results, retries, dead, nil
}

func parseReceiptResponse(batchID string, value interface{}, operationErr error) (Receipt, error) {
	parts, ok := value.([]interface{})
	if !ok || len(parts) == 0 {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: malformed script response", operationErr)
	}
	statusText, err := scriptString(parts[0])
	if err != nil {
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: status: %v", operationErr, err)
	}
	status := ReceiptStatus(statusText)
	receipt := Receipt{Status: status, BatchID: batchID}
	switch status {
	case ReceiptApplied, ReceiptDuplicate, ReceiptConflict, ReceiptStale:
		if len(parts) < 2 {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: receipt is missing", operationErr)
		}
		raw, rawErr := scriptString(parts[1])
		if rawErr != nil || raw == "" {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: receipt response: %v", operationErr, rawErr)
		}
		stored, decodeErr := decodeStoredReceipt(raw)
		if decodeErr != nil {
			return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: receipt: %v", operationErr, decodeErr)
		}
		receipt = publicReceipt(stored, status)
	case ReceiptNotFound, ReceiptNotDue:
		// These statuses intentionally carry no stored receipt.
	case ReceiptInvalid:
		// The error text is handled below.
	default:
		return Receipt{Status: ReceiptInvalid, BatchID: batchID}, fmt.Errorf("%w: unknown script status %q", operationErr, status)
	}
	if status == ReceiptInvalid {
		message := "script rejected command"
		if len(parts) > 1 {
			message, _ = scriptString(parts[1])
		}
		return receipt, fmt.Errorf("%w: %s", operationErr, message)
	}
	return receipt, nil
}
